Integration test fails randomly in ci runs

For some reason unbeknown to me, the integration test fails from time to time. See the following [run](https://github.com/MindTooth/fish-code/runs/3134354844?check_suite_focus=true).

From what I can gather, it seems to crash on reading JSON data supplied from requests.  But also complains about something SQLAlchemy related too.  Initial thought, is that they are related.

From the logs in sonar: 
```            
 response = client.get(
                "/projects/1/jobs/1/csv",
                follow_redirects=True,
            )
>           assert response.status_code == 200
E           assert 500 == 200
E            +  where 500 = <Response streamed [500 INTERNAL SERVER ERROR]>.status_code
```
Did you check if this endpoint is working properly?

--- a/src/ui/projects/api.py
+++ b/src/ui/projects/api.py
@@ -3,6 +3,7 @@ import base64
 import functools
 import json
 import logging
+from http import HTTPStatus
 from typing import (
     Any,
     Callable,
@@ -111,7 +112,7 @@ class Client:
                 def wrapper_call(*args, **kwargs) -> Result[requests.Response]:  # type: ignore
                     try:
                         response = request(*args, **kwargs)  # type: ignore
-                    except requests.ConnectionError as e:
+                    except (requests.ConnectionError, requests.Timeout) as e:
                         logger.warning(
                             f"ConnectionError: {request.__name__} - "
                             f"{request.__dict__} - {repr(e)}"
@@ -345,18 +346,24 @@ class Client:
                     Single `Job` from `core`.
         """
         result_project = self.get(f"{self._endpoint}/projects/{project_id}")
+        if not isinstance(result_project, requests.Response) or (
+            isinstance(result_project, requests.Response)
+            and result_project.status_code == HTTPStatus.NOT_FOUND
+        ):
+            return None
+
         result_job = self.get(
             f"{self._endpoint}/projects/{project_id}/jobs/{job_id}"
         )
-
-        if isinstance(result_project, requests.Response) and isinstance(
-            result_job, requests.Response
+        if not isinstance(result_job, requests.Response) or (
+            isinstance(result_job, requests.Response)
+            and result_job.status_code == HTTPStatus.NOT_FOUND
         ):
-            return Job.from_dict(
-                result_job.json(), project_id, result_project.json()["name"]  # type: ignore
-            )
+            return None
 
-        return None
+        return Job.from_dict(
+            result_job.json(), project_id, result_project.json()["name"]
+        )
 
     def post_job(self, project_id: int, job: Job) -> Optional[int]:
         """Post a single job to the endpoint.

--- a/tests/integration/test_services.py
+++ b/tests/integration/test_services.py
@@ -1,6 +1,7 @@
 """Integration test between model and services."""
 import logging
 import time
+from json import JSONDecodeError
 from pathlib import Path
 from typing import Dict, List
 
@@ -64,9 +65,19 @@ def test_processing_and_scheduler():
 
     assert response_finished.status_code == 200
 
-    while requests.get(url).json()["_status"] != "Done":
+    status = ""
+    while status != "Done":
         logger.info("Waiting for job to finish.")
-        time.sleep(15)
+        time.sleep(5)
+        res = requests.get(url)
+
+        try:
+            status = res.json().get("_status", "")
+        except JSONDecodeError as e:
+            logger.warning(
+                f"Response not valid json, error: {repr(e)}, response: {repr(res)}"
+            )
+            continue
 
     job_data = requests.get(url).json()
     print(job_data)

--- a/tests/unit/ui/test_ui.py
+++ b/tests/unit/ui/test_ui.py
@@ -1,6 +1,7 @@
 """Unit testing ui Flask app and api Client."""
 import json
 from dataclasses import asdict
+from http import HTTPStatus
 
 import pytest
 import requests
@@ -8,7 +9,7 @@ from werkzeug.exceptions import HTTPException
 
 from ui.main import create_app
 from ui.projects.api import Client
-from ui.projects.model import Object, ProjectBare
+from ui.projects.model import Job, Object, ProjectBare
 
 TEST_API_URI = "mock://testing"
 TEST_API_URI_ERROR = "mock://testingerrors"
@@ -163,6 +164,21 @@ def mock_client(requests_mock):
     requests_mock.get(f"{TEST_API_URI}/projects/1/jobs/1", json=job)
     requests_mock.get(f"{TEST_API_URI}/projects/1/jobs/2", json=job_done)
 
+    # mock Client.get_job - broken
+    requests_mock.get(
+        f"{TEST_API_URI}/projects/13", json={}, status_code=HTTPStatus.NOT_FOUND
+    )
+    requests_mock.get(
+        f"{TEST_API_URI}/projects/1/jobs/31",
+        json={},
+        status_code=HTTPStatus.NOT_FOUND,
+    )
+    requests_mock.get(
+        f"{TEST_API_URI}/projects/13/jobs/31",
+        json={},
+        status_code=HTTPStatus.NOT_FOUND,
+    )
+
     # mock core object preview endpoint
     requests_mock.get(f"{TEST_API_URI}/objects/1/preview", json={})
 
@@ -216,6 +232,27 @@ def test_client_get_project(mock_client):
     assert response is not None
     assert isinstance(response, ProjectBare)
     assert response.name == "Test name"
+
+
+def test_client_get_job(mock_client):
+    """Unit test the Client."""
+    client = Client(TEST_API_URI)
+
+    response = client.get_job(1, 1)
+    assert response is not None
+    assert isinstance(response, Job)
+    assert response.name == "string"
+
+
+def test_client_get_noexisting_job(mock_client):
+    """Unit test the Client."""
+    client = Client(TEST_API_URI)
+
+    response = client.get_job(1, 31)
+    assert response is None
+
+    response = client.get_job(13, 31)
+    assert response is None
 
 
 def test_client_get_objects(mock_client) -> None:

